### ISSUE TITLE
Add basic throttling wrappers

### ISSUE THREAD

**Issue by dahall** [octokit/octokit.net#1792]
Throttling to avoid abuse and rate limit exceptions follows a consistent pattern. It would be very valuable to build it into each method call or provide extension methods to assist developers. I have included the ones I've built using my limited understanding.
```cs
private static async Task<T> Throttle<T>(this GitHubClient client, Func<GitHubClient, Task<T>> method, CancellationToken cancellationToken, bool throwIfRateLimitExceeded = false)
{
   while (!cancellationToken.IsCancellationRequested)
   {
      try
      {
         var ret = await method(client).ConfigureAwait(false);
         await Task.Delay(1000, cancellationToken).ConfigureAwait(false);
         return ret;
      }
      catch (AbuseException abuse)
      {
         await Task.Delay(TimeSpan.FromSeconds(abuse.RetryAfterSeconds.GetValueOrDefault(30)), cancellationToken).ConfigureAwait(false);
      }
      catch (RateLimitExceededException limit)
      {
         if (throwIfRateLimitExceeded) throw;
         await Task.Delay(DateTimeOffset.Now - limit.Reset, cancellationToken).ConfigureAwait(false);
      }
   }
   throw new InvalidOperationException();
}

private static async Task Throttle(this GitHubClient client, Func<GitHubClient, Task> method, CancellationToken cancellationToken, bool throwIfRateLimitExceeded = false)
{
   while (!cancellationToken.IsCancellationRequested)
   {
      try
      {
         await method(client).ConfigureAwait(false);
         await Task.Delay(1000, cancellationToken).ConfigureAwait(false);
      }
      catch (AbuseException abuse)
      {
         await Task.Delay(TimeSpan.FromSeconds(abuse.RetryAfterSeconds.GetValueOrDefault(30)), cancellationToken).ConfigureAwait(false);
      }
      catch (RateLimitExceededException limit)
      {
         if (throwIfRateLimitExceeded) throw;
         await Task.Delay(DateTimeOffset.Now - limit.Reset, cancellationToken).ConfigureAwait(false);
      }
   }
   throw new InvalidOperationException();
}
```

**Comment by ryangribble** [comment]
I'm a bit hesitant to include things like this in the core library, as we try to keep it as much of a lightweight wrapper around the upstream GitHub api as possible.  This is why we don't have "composite" actions (involving multiple API calls) or built in caching.  I think it's difficult to match the different use cases and expectations of users when you start doing things like this.  Particularly in this example, it could end up "sleeping" for almost an hour.  (I'm also questioning why it has a hardcoded 1 second sleep, after any response is received?)   

Thanks for sharing though, and we can keep this here as an example for anyone wanting to know how they could wrap calls to automatically handle rate limit or abuse retries.  One thing I've been thinking about, is creating a separate NuGet package for `Octokit.Extensions` that could include more of these additional helpers, without having them in the "officially supported" library...

**Comment by shiftkey** [comment]
For reference, I wrote [a similar thing](https://github.com/shiftkey/paket-upgrader/blob/master/PaketUpgrader/Wrap.cs) a while ago in a hurry that was designed for batch tasks and doing as much work as possible with the available quota:

 - CLI-focused - displays messages around available quota while working
 - fail-fast - once you've exceeded your quota, show when you can start over and exit

It's not as exhaustive as the initial example, but also doesn't wait for your quota to reset - just wanted to illustrate that there are different approaches here around rate-limiting.

**Comment by dahall** [comment]
@ryangribble It makes sense to not include in the base library for your reasons. However, by having the rate limit information return in the header, it is very easy for first time users to almost immediately hit the limit and then have to wade through a lot of documentation to figure out how to prevent it. Thus my recommendation. The "1 second" is from some of the documentation as a recommendation. If it can be removed, I'll remove it from the post so others don't include it if they reuse it.

**Comment by dahall** [comment]
Another thought, given the challenge of using the library for multiple calls or "composite" calls, may be to develop a queue into which the user can submit calls and have the rate limiting done for them. If that is something you'd consider, I'm willing to do the the initial development work on it. I'm just looking for ways to keep others from hitting the problems I did and having to spend all the time wading through the docs to get a workable solution.

**Comment by ryangribble** [comment]
> @ryangribble It makes sense to not include in the base library for your reasons. However, by having the rate limit information return in the header, it is very easy for first time users to almost immediately hit the limit and then have to wade through a lot of documentation to figure out how to prevent it. Thus my recommendation. 

My concern is that if there was a "magic" wrapper that helped people wait out the rate limit then they are even less likely to know what they are doing (or the potential impact they may be causing).  We'd then have people thinking that "octokit took 30 minutes to execute a simple query" and things like that 😮 

I'd definitely be for improving any documentation that wasn't clear in how things work, although I did think we specifically called out how rate limiting works and highlighted the fact that you need to use authenticated requests if you want to be given the more generous rate limits from GitHub.

> The "1 second" is from some of the documentation as a recommendation. If it can be removed, I'll remove it from the post so others don't include it if they reuse it.

Did you have a link to this?  The implementation as provided always waits 1 second AFTER a successful response... I think you should only wait if an abuse or rate limit was hit, which your code already does. 
 Im not sure it makes sense to always delay 1 second AFTER a successful response

> Another thought, given the challenge of using the library for multiple calls or "composite" calls, may be to develop a queue into which the user can submit calls and have the rate limiting done for them. If that is something you'd consider, I'm willing to do the the initial development work on it. I'm just looking for ways to keep others from hitting the problems I did and having to spend all the time wading through the docs to get a workable solution.

Again, I think things like this really fall out of the domain of an API wrapper... there are so many different ways people may want to implement queuing, rate limit handling or caching, we don't really want the additional complexity in this library.  If you wanted to put together your own extension package, or provide a gist with the code samples, I'd be happy to take a look at the approach, but it wouldn't be something we would include in the core library :+1:

**Comment by mirsaeedi** [comment]
I just saw this thread. I got the idea from @ryangribble and I solved the rate limiting issue and the retry policy by using Polly and DelegatingHandler of HttpClient. I created a [github gist](https://gist.github.com/mirsaeedi/e11f903b384861941624a49fa34ef3d5) to share my code with other folks.

Response and ApiInfoParser classes and BuildResponse method are copied exactly from the octokit's source code. I had to copy them because they weren't defined as public classes and methods.
